source/parser.hera

```diff
@@ -7951,33 +7952,102 @@ Init
         let   [ , arg ]   = s.children
 
         let i = 0, l = body.length
-        const children = [ws]
+
+        const refDec = []
+        const children = [ws, refDec]
+
+        let usingRef = null
 
         for (i = 0; i < l; i++) {
           const step = body[i]
           const [leadingComment, pipe, trailingComment, expr] = step
           const returns = pipe.token === "||>"
+          let ref, result,
+            returning = returns ? arg : null
 
           if (pipe.token === "|>=") {
+            let initRef
             if (i === 0) {
-              // TODO: assignment node
-              children.push(arg, " = ")
+              switch(arg.type) {
+                case "MemberExpression":
+                  // If there is only a single access then we don't need a ref
+                  if (arg.children.length <= 2) break
+                case "CallExpression":
+                  const access = arg.children.pop()
+                  usingRef = module.needsRef({}) // hacky: using this like a "createRef"
+                  initRef = {
+                    type: "AssignmentExpression",
+                    children: [usingRef, " = ", arg, ";"],
+                  }
+
+                  arg = {
+                    type: "MemberExpression",
+                    children: [usingRef, access]
+                  }
+
+                  break;
+              }
+
+              // remove refDec from children and put it inside lhs of assignment
+              // TODO: should be attached to node to be hoisted
+              children.pop()
+              // assignment node
+              const lhs = [[
+                [refDec, initRef],
+                arg,
+                [],
+                {token: "=", children: [" = "]}
+              ]];
+
+              Object.assign(s, {
+                type: "AssignmentExpression",
+                children: [ lhs, children ],
+                names: null,
+                lhs,
+                assigned: arg,
+                exp: children,
+              })
+
+              // Clone so that the same node isn't on the left and right because splice manipulation
+              // moves things around and can cause a loop in the graph
+              arg = structuredClone(arg)
```

Comment:
I'm worried what happens with parent pointers here. Does structuredClone end ip traversing the entire AST? I wonder if there's a way to avoid this... Can you just clone what you end up changing?

Reply:
I could null out the parent pointers, then `structuredClone`. They'll get added back at the end of the method.

Reply:
Or write our own simple clone method for ASTs... Not sure which is faster.

Reply:
Our own clone could be handy. It should keep refs the same for example.